plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Core/Type.hs

```diff
@@ -31,6 +31,22 @@ import           PlutusCore.MkPlc
 import qualified PlutusCore.Name                        as TPLC
 import           Universe
 
+{- Note [Term constructor ordering and numbers]
+Ordering of constructors has a small but real effect on efficiency.
```

Comment:
If we're doing this for the `Term` type, I suggest we also do it for `StepKind` and also `CekValue` in `Cek/Internal.hs`, and maybe also `Frame` (although that looks like a pretty sensible order already).   I'm not sure how often the various `CekValue` types are seen, but we can find out.

Reply:
Actually the `CekValue` types are used in different ways, so it's tricky to tell how important the frequencies are.  The only place we dispatch on the full range is in `dischargeCekValue`, which is irrelevant since it shoud only ever be called once per program.  Apart from that `VLamAbs` and `VBuiltin` are used in two cases of `forceEvaluate` (when you're applying a lambda or a builtin to an argument and `VDelay` and `VBuiltin` are used in `forceEvaluate` (when you're applying `force` to something you've just reduced to a value).  I think `VCon` is only ever used is in `asConstant`  when function arguments are passed to the builtin evaluation machinery, and I have no idea if the tag of the constructor matters there.

Reply:
Based on the figures for the `Term` type it looks as if a sensible order would be `VLamAbs`, `VDelay`, `VCon`, `VBuiltin`.  Not sure about the order of the last two, if it even matters.  

Also, does it matter what order we do matches in, or are they translated to something based on the constructor order?

Reply:
Hmm, good idea about `CekValue`. I'll give that a try, and I agree with your analysis.

I *believe* the order of matches doesn't matter, it's just the order of the constructors.